src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,187 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode 
+    {
+        internal sealed class MoveFileCodeAction : CodeAction
+        {
+            private readonly State _state;
+            private readonly ImmutableArray<string> _newfolders;
+
+            public override string Title
+                => _newfolders.Length > 0 
+                ? string.Format(FeaturesResources.Move_file_to_0_folder, string.Join(PathUtilities.DirectorySeparatorStr, _newfolders)) 
+                : FeaturesResources.Move_file_to_project_root_folder;
+
+            public MoveFileCodeAction(State state, ImmutableArray<string> newFolders)
+            {
+                _state = state;
+                _newfolders = newFolders;
+            }
+
+            internal override bool PerformFinalApplicabilityCheck => true;
+
+            internal override bool IsApplicable(Workspace workspace)
+            {
+                // Due to some existing issue, move file action is not available for CPS projects.
+                return workspace.CanRenameFilesDuringCodeActions(workspace.CurrentSolution.GetDocument(_state.OriginalDocumentId).Project);
+            }
+
+            protected override Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+                => MoveFileToMatchNamespaceAsync(cancellationToken);
+
+            private async Task<IEnumerable<CodeActionOperation>> MoveFileToMatchNamespaceAsync(CancellationToken cancellationToken)
+            {
+                var id = _state.OriginalDocumentId;
+                var solution = _state.Solution;
+                var document = solution.GetDocument(id);
+                var newDocumentId = DocumentId.CreateNewId(document.Project.Id, document.Name);
+
+                solution = solution.RemoveDocument(id);
+
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                solution = solution.AddDocument(newDocumentId, document.Name, text, folders: _newfolders);
+
+                return ImmutableArray.Create<CodeActionOperation>(
+                    new ApplyChangesOperation(solution),
+                    new OpenDocumentOperation(newDocumentId, activateIfAlreadyOpen: true));
+            }
+            
+            public static ImmutableArray<MoveFileCodeAction> Create(State state)
+            {
+                // Since all documents have identical folder structure, we can do the computation on any of them.
+                var document = state.Solution.GetDocument(state.DocumentIds.First());
```

Comment:
it's a little unclear why we have multiple DocumentIds for a MoveFileCodeAction.  I would think we'd only have one.

Reply:
It could be a document in a multi-targeting project. It's not needed for this action, but "change namespace" action depends on that info, and `State` is shared between those two actions.

Reply:
The same comment from earlier applies.  the 'DocumentIds' field is not doc'ed at all, making this unclear.

Reply:
Added comments